### DISCUSSION TITLE
API update question

### DISCUSSION THREAD

**Discussion by manderson88** [iTwin/itwinjs-core#4439] [Q&A]
I am working on (finally) updating an application we have from the 2.x api to the 3.x api.  We are using the event subscription as shown in the example for working with changes.  I cannot find anything in the new api that is similar.  So how can I subscribe to imodelhub changes?

TIA,
marka

**Comment by pmconne** [comment]
Be specific about what API and/or example you are referring to please.

**Reply by swwilson-bsi** [reply]
Mark, do you mean iModelHub events (ChangeSetPostPushEvent, etc.)? I would like to learn how to do this, too.
In version 2.x, IModelHost.iModelClient was an `IModelClient` from `@bentley/imodelhub-client`. Its `events` property gave you access to the IMH event subscription system.
In version 3.x, the client is-a `IModelsClient` from `@itwin/imodels-client-authoring`. It has no events API. There is no mention of iModelHub events in itwinjs core that I can see.
There appears to be a new webhook-based event system. [api](https://prod-apim-developer-portal-eus.azurewebsites.net/apis/webhooks/events/), [tutorial](https://prod-apim-developer-portal-eus.azurewebsites.net/tutorials/react-to-events-using-webhooks/). That is all for frontend React code.

**Reply by manderson88** [reply]
I am based on this example: https://github.com/iTwin/agent-starter  It used the IModelHubClient EventHandler object to subscribe to the update events. 
    // Create iModelHub event subscription
    const eventTypes = [
      IModelHubEventType.ChangeSetPostPushEvent,
    ];
 this.hubSubscription = await this.hubClient.events.subscriptions.create(ctx, iModelId, eventTypes);
    Logger.logInfo(loggerCategory, `Event subscription "${this.hubSubscription.wsgId}" created in iModelHub.`);
    // Define event listener
    const listener = async (event: ChangeSetPostPushEvent) => {}

 Yes, I have also seen the webhooks api example and not sure if it would be able to handle the processing that I need.  
The current application runs and consumes the updates to put the changes into our backend database.  This allows us to support out legacy applications without wholesale updating of our system.  
Thanks for any guidance,

marka

**Reply by swwilson-bsi** [reply]
IModelHubClient is still there, in @bentley/imodelhub-client, and agent-starter still works.
Have you had problems using IModelHubClient?

**Reply by manderson88** [reply]
I am using the 2.x and not having any issues.  It has been stressed(our security guys push to be on latest of things) to get sync'd up with Bentley on the version so I have been moving to 3.x api.  I am not seeing this class in the 3.x docs.  Is there a replacement in the works or should I stay on 2.x until I come up with alternate.

thanks,
marka

**Reply by swwilson-bsi** [reply]
IModelHubClient is in @bentley/imodelhub-client. Is that package not part of 3.x? @pmconne ?

**Reply by pmconne** [reply]
A major part of 3.0 was refactoring the client packages, including @bentley/imodelhub-client.
I was not heavily involved in that process and I can find no information in 3.0.0.md change notes regarding this package.
@calebmshafer can help.

**Reply by pmconne** [reply]
@robertas-kerpys @Animagne

**Reply by robertas-kerpys** [reply]
IModelsClient is a client wrapper for [iModels API](https://developer.bentley.com/apis/imodels/). iTwin Platform events were exposed as [webhooks](https://developer.bentley.com/apis/webhooks/). Webhooks can only be consumed by applications which have publicly accessible URL which could be called upon webhook event. 

If your application is a client side app then you would need to build a helper service which would be able to subscribe to webhook events and provide polling interface for a client side application.
If your application can have a publicly accessible endpoint exposed then it is a smaller change to get those events.

Please refer to the iTwin platform [webhook tutorial](https://developer.bentley.com/tutorials/react-to-events-using-webhooks/) on how to use it in your application.
